src/multiplexed-session.ts

```diff
@@ -0,0 +1,241 @@
+/*!
+ * Copyright 2024 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EventEmitter} from 'events';
+import {Database} from './database';
+import {Session} from './session';
+import {
+  isDatabaseNotFoundError,
+  isInstanceNotFoundError,
+  isCreateSessionPermissionError,
+  isDefaultCredentialsNotSetError,
+  isProjectIdNotSetInEnvironmentError,
+} from './helper';
+import {GetSessionCallback} from './session-pool';
+
+/**
+ * Interface for implementing multiplexed session logic, it should extend the
+ * {@link https://nodejs.org/api/events.html|EventEmitter} class
+ *
+ * @interface MultiplexedSessionInterface
+ * @extends external:{@link https://nodejs.org/api/events.html|EventEmitter}
+ *
+ * @constructs MultiplexedSessionInterface
+ * @param {Database} database The database to create a multiplexed session for.
+ */
+export interface MultiplexedSessionInterface {
+  /**
+   * When called creates a multiplexed session.
+   *
+   * @name MultiplexedSessionInterface#createSession
+   */
+  createSession(): void;
+
+  /**
+   * When called returns a multiplexed session.
+   *
+   * @name MultiplexedSessionInterface#getSession
+   * @param {GetSessionCallback} callback The callback function.
+   */
+  getSession(callback: GetSessionCallback): void;
+}
+
+/**
+ * Class used to manage connections to Spanner using multiplexed session.
+ *
+ * **You don't need to use this class directly, connections will be handled for
+ * you.**
+ *
+ * @class
+ * @extends {EventEmitter}
+ */
+export class MultiplexedSession
+  extends EventEmitter
+  implements MultiplexedSessionInterface
+{
+  database: Database;
+  // frequency to create new mux session
+  refreshRate: number;
+  _multiplexedSession: Session | null;
+  _pingHandle!: NodeJS.Timer;
+  _refreshHandle!: NodeJS.Timer;
+  constructor(database: Database) {
+    super();
+    this.database = database;
+    // default frequency is 7 days
+    this.refreshRate = 7;
+    this._multiplexedSession = null;
+  }
+
+  /**
+   * Creates a new multiplexed session and manages its maintenance.
+   *
+   * This method initiates the session creation process by calling the `_createSession` method, which returns a Promise.
+   */
+  createSession(): void {
```

Comment:
A thought , can we make this method to return `Promise<void>` . This will allow us to use await with this method and if we dont want to await we can simply call this without await.